### PR TITLE
flux-kvs: output --help if not in an instance

### LINKED ISSUE #3491: `--help` fails for some commands without a running instance
@Larofeticus pointed out that some commands when passed `--help` fail unless you are in an active instance.  We should look into making all of the command's `--help` work even outside of a running Flux instance.  I think this mainly involves moving `flux_open` calls until after argument parsing.

### PR DESCRIPTION
**Pull request by chu11**
Problem: Sometimes --help will not output help output if the
user is not running in a flux instance.

Solution: flux-kvs would flux_open() before all options were
parsed in subcommands.  Instead every subcommand should call
flux_open().  Each subcommand knows when it has completed all
option parsing.

Fixes #3491

### CODE REVIEW

**Comment by chu11** [comment]
note, I didn't add a test since this is fix is specific to the case when you're not under a flux instance, so it wouldn't work under the `t4000` tests.  I could add a `t4001-issues-not-under-flux`, but wasn't sure if it was worthwhile for such a unique corner case.

I also did a mini-audit of other flux commands and it seems this issue only existed in `flux-kvs`.

**Comment by chu11** [comment]
@mergifyio rebase

**Comment by grondo** [comment]
> Or perhaps I could stick it in t0001-basic or a similar place?

Sorry, somehow missed this comment. It does seem like a simple test could go into `t0001-basic.t`.. but I wouldn't say it is a requirement if you'd rather not.

Maybe we could add a test in the future that spins through all commands and ensures `--help` works without connecting to flux...

**Comment by chu11** [comment]
> Sorry, somehow missed this comment. It does seem like a simple test could go into t0001-basic.t.. but I wouldn't say it is a requirement if you'd rather not.

I'll do that then, adding a bunch of `--help` tests in `t0001-basic`.

**Comment by chu11** [comment]
pushed an extra commit on top to `t0001-basic` to check that `--help` works w/o being in an instance.  Only tested the C commands, since we're not in an instance, so `flux python` isn't available anyways.

File: t/t0001-basic.t [hunk 1]
```diff
@@ -409,6 +409,24 @@ test_expect_success 'no unit tests built with libtool wrapper' '
 	test_must_fail grep -q "\.libs" test_dirs
 '
 
+test_expect_success 'command --help works outside of flux instance' '
+        flux --help 2>&1 |  grep -i usage &&
+        flux-R --help 2>&1 |  grep -i usage &&
```
Comment (grondo) [inline t/t0001-basic.t:414]: Why are you running the `flux-CMD` version of the commands instead of `flux CMD`, as users would? Just curious.

Also, it might be easier to maintain this list going forward to have the commands to test in a list, then loop through them running `$CMD --help` each time. Then in the future if we have a way for `flux(1)` to dump a list of all available commands (e.g. for use with command completion), we can switch to that.

Comment (chu11) [reply t/t0001-basic.t:414]: oh haha, i associated "not in an instance" with "flux command won't search paths", but that's not true.  I'll fixup and add the python commands too.

**Comment by chu11** [comment]
re-pushed, with fixes per @grondo comment above

File: t/t0001-basic.t [hunk 2]
```diff
@@ -409,6 +409,42 @@ test_expect_success 'no unit tests built with libtool wrapper' '
...
+
+KVS_SUBCMDS="\
+namespace \
+eventlog \
+"
+
+test_expect_success 'command --help works outside of flux instance' '
+        flux --help 2>&1 | grep -i usage &&
+        for cmd in ${CMDS}; do
+            flux ${cmd} --help 2>&1 | grep -i usage
```
Comment (garlick) [inline t/t0001-basic.t:442]: I think if one of the commands in the for loop fails, the for loop will still succeed.  See `test_valid()` in `t2200-ingest.t` for one way to do it.

Comment (chu11) [reply t/t0001-basic.t:442]: I tested by doing things like:

```
--- a/t/t0001-basic.t
+++ b/t/t0001-basic.t
@@ -429,6 +429,7 @@ queue \
 resource \
 start \
 terminus \
+foobar \
 "
```

so I thought it was safe.  But perhaps some corner cases I'm missing.  I'll to the `test_valid()` technique you did in `t2200-job-ingest`

**Comment by chu11** [comment]
re-pushed with tweak per @garlick comments above

File: t/t0001-basic.t (hunk 2 above)
Comment (garlick) [reply t/t0001-basic.t]: Right I think the last command's return code _does_ affect the result, just not the ones before that.  Your changes look good to me.

**Review by garlick — APPROVED** [review]
Looks good, thanks!
